packages/app/lib/notifications.ts

```diff
@@ -69,3 +74,62 @@ export const connectNotifications = async () => {
     return false;
   }
 };
+
+const channelIdFromNotification = (notif: Notifications.Notification) => {
+  if (notif.request.trigger.type !== 'push') {
+    return null;
+  }
+  const out = notif.request.trigger.payload?.channelId;
+  if (typeof out !== 'string') {
+    return null;
+  }
+  return out;
+};
+
+/**
+ * Imprecise method to sync internal unreads with presented notifications.
+ * We should move to a serverside badge + dismiss notification system, and remove this.
+ */
+async function updatePresentedNotifications() {
+  const presentedNotifs = await Notifications.getPresentedNotificationsAsync();
+  const allChannelIds = new Set(
+    compact(presentedNotifs.map(channelIdFromNotification))
+  );
+
+  const fullyReadChannels = new Set<string>();
+  for await (const channelId of allChannelIds) {
+    const channel = await db.getChannelWithRelations({ id: channelId });
+    if (channel?.unread?.count === 0) {
+      fullyReadChannels.add(channelId);
+    }
+  }
+
+  const notificationsToDelete = presentedNotifs.filter((notif) => {
+    const cId = channelIdFromNotification(notif);
+    // also delete notifications that have no channel id to avoid stuck notifs
+    return cId == null || fullyReadChannels.has(cId);
+  });
+
+  await Promise.all(
+    notificationsToDelete.map(async (notif) => {
+      await Notifications.dismissNotificationAsync(notif.request.identifier);
+    })
+  );
+
+  const stillPresented = await Notifications.getPresentedNotificationsAsync();
+  await Notifications.setBadgeCountAsync(stillPresented.length);
+}
+
+export function useUpdatePresentedNotifications() {
+  const query = store.useUnreadsCount();
+
+  // `useUnreadsCount` updates with a lot of false positives - debounce so we
+  // don't run the updater too frequently
+  const debouncedQueryKey = useDebouncedValue(query.dataUpdatedAt, 500);
```

Comment:
Since the `query.data` is just `number | undefined`, why not use that number as the key?

Reply:
I was thinking about this case: This query's data could remain the same but badge could need a change if there is an equal number of +/- read/unreads in different rooms in one transaction, causing no change in the total unread count. (_If that's the case, wouldn't the badge also reflect no change in value?_ Not necessarily, because the badge only represents non-muted channels - I thought useUnreadsCount contains muted channels.)

I'll leave a comment inline if that makes sense

Reply:
You're right that it does -- fwiw `useUnreadsCount` does not appeared to be use anywhere else besides here, so feel free to modify if helpful.